crates/edr_napi/src/provider.rs

```diff
@@ -0,0 +1,58 @@
+mod config;
+
+use std::sync::Arc;
+
+use napi::{tokio::runtime, Env, JsObject, Status};
+use napi_derive::napi;
+
+use self::config::ProviderConfig;
+
+/// A JSON-RPC provider for Ethereum.
+#[napi]
+pub struct Provider {
+    provider: Arc<edr_provider::Provider>,
+}
+
+#[napi]
+impl Provider {
+    #[doc = "Constructs a new provider with the provided configuration."]
+    #[napi(ts_return_type = "Promise<Provider>")]
+    pub fn with_config(env: Env, config: ProviderConfig) -> napi::Result<JsObject> {
+        let config = edr_provider::ProviderConfig::try_from(config)?;
+        let runtime = runtime::Handle::current();
+
+        let (deferred, promise) = env.create_deferred()?;
+        runtime.clone().spawn(async move {
+            let result = edr_provider::Provider::new(&runtime, &config)
+                .await
+                .map_or_else(
+                    |error| Err(napi::Error::new(Status::GenericFailure, error.to_string())),
+                    |provider| {
+                        Ok(Provider {
+                            provider: Arc::new(provider),
+                        })
+                    },
+                );
+
+            deferred.resolve(|_env| result);
+        });
+
+        Ok(promise)
+    }
+
+    #[doc = "Handles a JSON-RPC request and returns a JSON-RPC response."]
+    #[napi]
+    pub async fn handle_request(&self, json_request: String) -> napi::Result<String> {
+        let request = serde_json::from_str(&json_request)
+            .map_err(|e| napi::Error::new(Status::InvalidArg, e.to_string()))?;
```

Comment:
Should we include the request in the error or log it? IIRC `serde_json::Error` doesn't include the source string which will make debugging from JS hard.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/4578/commits/0f392206571870cb24275b3c15236ddf39c5b99f